### PR TITLE
Bux 396/endpoint for creating contact

### PR DESCRIPTION
**Pull request by Nazarii-4chain**
Added contact endpoint

# Pull Request Checklist

- [x] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/BuxOrg/bux-server/blob/main/.github/CODE_STANDARDS.md)
- [x] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/BuxOrg/bux-server/blob/main/.github/CODE_OF_CONDUCT.md)
- [x] 🏠 I tested my changes locally.
- [x] ✅ I have provided tests for my changes.
- [x] 📝 I have used conventional commits.
- [x] 📗 I have updated any related documentation.
- [x] 💾 PR was issued based on the Github or Jira issue.

### CODE REVIEW

File: engine/model_contact.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+	return contact, nil
+}
+
+func getContact(ctx context.Context, fullName, paymailAddress, senderPubKey string, opts ...ModelOps) (*Contact, error) {
+	contact, err := newContact(fullName, paymailAddress, senderPubKey, opts...)
+	if err != nil {
+		return nil, err
+	}
+	contact.ID = ""
+	conditions := map[string]interface{}{
```
Comment (arkadiuszos4chain) [inline engine/model_contact.go:56]: We should search for contact by sender xpub and contact paymail

**Review by arkadiuszos4chain — CHANGES_REQUESTED** [review]

File: actions/contacts/create.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	paymail := params.GetString("paymail")
+	pubKey := params.GetString("pubKey")
+	metadata := params.GetJSON("metadata")
+
+	opts := c.Services.SpvWalletEngine.DefaultModelOptions()
+	if metadata != nil {
+		opts = append(opts, engine.WithMetadatas(metadata))
+	}
+
+	contact, err := c.Services.SpvWalletEngine.NewContact(req.Context(), fullName, paymail, pubKey, opts...)
```
[the remark this answers is not in the document]
Comment (Nazarii-4chain) [reply actions/contacts/create.go]: Changed

File: engine/model_contact.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+		if errors.Is(err, datastore.ErrNoResults) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	return contact, nil
+}
+
+func (c *Contact) getContactPaymailCapability(ctx context.Context) (*paymail.CapabilitiesPayload, error) {
```
[the remark this answers is not in the document]
Comment (Nazarii-4chain) [reply engine/model_contact.go]: Moved

File: engine/model_contact.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+		if errors.Is(err, cachestore.ErrKeyNotFound) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	return capabilities, nil
+}
+
+func (c *Contact) getPubKeyFromPki(pkiUrl string) (string, error) {
```
[the remark this answers is not in the document]
Comment (Nazarii-4chain) [reply engine/model_contact.go]: Moved

File: engine/model_contact.go [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+func (c *Contact) BeforeCreating(_ context.Context) (err error) {
+	c.Client().Logger().Debug().
+		Str("contactID", c.ID).
+		Msgf("starting: %s BeforeCreate hook...", c.Name())
+
+	if c.ID == "" {
+		return ErrMissingContactID
+	}
+
+	if len(c.FullName) == 0 {
```
[the remark this answers is not in the document]
Comment (Nazarii-4chain) [reply engine/model_contact.go]: Changed

File: engine/model_contact.go [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+	}
+
+	c.Client().Logger().Debug().
+		Str("contactID", c.ID).
+		Msgf("end: %s BeforeCreate hook", c.Name())
+	return
+}
+
+// AfterCreated will fire after the model is created in the Datastore
+func (c *Contact) AfterCreated(_ context.Context) error {
```
[the remark this answers is not in the document]
Comment (Nazarii-4chain) [reply engine/model_contact.go:120]: It appears in others models too

File: engine/model_contact.go [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+// migrateMySQL is specific migration SQL for MySQL
+func (c *Contact) migrateMySQL(client datastore.ClientInterface, tableName string) error {
+	idxName := "idx_" + tableName + "_contacts"
+	idxExists, err := client.IndexExists(tableName, idxName)
+	if err != nil {
+		return err
+	}
+	if !idxExists {
+		tx := client.Execute("CREATE UNIQUE INDEX " + idxName + " ON `" + tableName + "` (full_name, paymail)")
```
[the remark this answers is not in the document]
Comment (Nazarii-4chain) [reply engine/model_contact.go]: Remove unique

File: engine/model_contact.go (hunk 1 above)
Comment (Nazarii-4chain) [reply engine/model_contact.go:57]: Changed

**Review by arkadiuszos4chain — APPROVED** [review]

**Pull request merged**